beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            });
+
+        // Set the Prometheus metrics for each validator (including missed_blocks_count=0)
+        validators_missed_blocks
+            .iter()
+            .for_each(|(validator_index, missed_blocks_count)| {
+                self.aggregatable_metric(validator_index.to_string().as_str(), |label| {
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL,
+                        &[label],
+                        u64_to_i64(*missed_blocks_count),
+                    );
+                });
+            });
+
+        // Increment the Prometheus metrics counter for each missed block
+        if let Some(i) = self.last_missed_block_validator {
+            self.aggregatable_metric(i.to_string().as_str(), |label| {
+                metrics::inc_counter_vec(&metrics::VALIDATOR_MONITOR_MISSED_BLOCKS_TOTAL, &[label]);
+            });
+            self.last_missed_block_validator = None;
+        };
+
+        // Prune missed blocks that are prior to last finalized epochs
+        let finalized_epoch = state.finalized_checkpoint().epoch;
+        self.missed_blocks
+            .retain(|(epoch, _, _)| *epoch > finalized_epoch);
+    }
+
+    /// Add missed non-finalized blocks for the monitored validators
+    fn add_validators_missed_blocks(&mut self, state: &BeaconState<T>) {
+        // Define range variables
+        let current_slot = state.slot();
+        let start_slot = current_slot.saturating_sub(T::slots_per_epoch()).as_u64();
+        let end_slot = current_slot.saturating_sub(MISSED_BLOCK_LAG_SLOTS).as_u64();
+
+        // As we are skipping the genesis slot, we can simply pass Hash256::zero() as the shuffling decision block
+        // cf. state.proposer_shuffling_decision_root_at_epoch implementation
+        let shuffling_decision_block = Hash256::zero();
+
+        // List of proposers per epoch from the beacon_proposer_cache
+        let mut proposers_per_epoch: Option<SmallVec<[usize; TYPICAL_SLOTS_PER_EPOCH]>> = None;
+
+        for (prev_slot, slot) in (start_slot..=end_slot).map(Slot::new).tuple_windows() {
+            // Condition for missed_block is defined such as block_root(slot) == block_root(slot - 1)
+            // where the proposer who missed the block is the proposer of the block at block_root(slot)
+            if let (Ok(block_root), Ok(prev_block_root)) =
+                (state.get_block_root(slot), state.get_block_root(prev_slot))
+            {
+                // Found missed block
+                if block_root == prev_block_root {
+                    let slot_epoch = slot.epoch(T::slots_per_epoch());
+                    let prev_slot_epoch = prev_slot.epoch(T::slots_per_epoch());
+
+                    if let Ok(shuffling_decision_block) = state
+                        .proposer_shuffling_decision_root_at_epoch(
+                            slot_epoch,
+                            shuffling_decision_block,
```

Comment:
I don't think it will make a practical difference, but I think we should use `block_root` rather than `Hash256::zero()` here.

`block_root` is the most correct value, but I don't think we'll ever hit the case where we need that value. I think it's worth doing though, especially since it means we can get rid of `let shuffling_decision_block = Hash256::zero();`.